Signal/src/util/OWSBackupImportJob.m

```diff
@@ -182,118 +201,174 @@ - (void)downloadAndProcessManifest:(OWSBackupJobCompletion)completion
 }
 
-- (void)downloadFilesFromCloud:(NSMutableArray<NSString *> *)recordNames completion:(OWSBackupJobCompletion)completion
+- (nullable NSArray<OWSBackupImportItem *> *)parseItems:(id)json key:(NSString *)key
 {
-    OWSAssert(recordNames.count > 0);
+    OWSAssert(json);
+    OWSAssert(key.length);
+
+    if (![json isKindOfClass:[NSDictionary class]]) {
+        OWSProdLogAndFail(@"%@ manifest has invalid data: %@.", self.logTag, key);
+        return nil;
+    }
+    NSArray *itemMaps = json[key];
+    if (![itemMaps isKindOfClass:[NSArray class]]) {
+        OWSProdLogAndFail(@"%@ manifest has invalid data: %@.", self.logTag, key);
+        return nil;
+    }
+    NSMutableArray<OWSBackupImportItem *> *items = [NSMutableArray new];
+    for (NSDictionary *itemMap in itemMaps) {
+        if (![itemMap isKindOfClass:[NSDictionary class]]) {
+            OWSProdLogAndFail(@"%@ manifest has invalid item: %@.", self.logTag, key);
+            return nil;
+        }
+        NSString *_Nullable recordName = itemMap[kOWSBackup_ManifestKey_RecordName];
+        NSString *_Nullable encryptionKeyString = itemMap[kOWSBackup_ManifestKey_EncryptionKey];
+        NSString *_Nullable relativeFilePath = itemMap[kOWSBackup_ManifestKey_RelativeFilePath];
+        NSNumber *_Nullable uncompressedDataLength = itemMap[kOWSBackup_ManifestKey_DataSize];
+        if (![recordName isKindOfClass:[NSString class]]) {
+            OWSProdLogAndFail(@"%@ manifest has invalid recordName: %@.", self.logTag, key);
+            return nil;
+        }
+        if (![encryptionKeyString isKindOfClass:[NSString class]]) {
+            OWSProdLogAndFail(@"%@ manifest has invalid encryptionKey: %@.", self.logTag, key);
+            return nil;
+        }
+        // relativeFilePath is an optional field.
+        if (relativeFilePath && ![relativeFilePath isKindOfClass:[NSString class]]) {
+            OWSProdLogAndFail(@"%@ manifest has invalid relativeFilePath: %@.", self.logTag, key);
+            return nil;
+        }
+        NSData *_Nullable encryptionKey = [NSData dataFromBase64String:encryptionKeyString];
+        if (!encryptionKey) {
+            OWSProdLogAndFail(@"%@ manifest has corrupt encryptionKey: %@.", self.logTag, key);
+            return nil;
+        }
+        // uncompressedDataLength is an optional field.
+        if (uncompressedDataLength && ![uncompressedDataLength isKindOfClass:[NSNumber class]]) {
+            OWSProdLogAndFail(@"%@ manifest has invalid uncompressedDataLength: %@.", self.logTag, key);
+            return nil;
+        }
+
+        OWSBackupImportItem *item = [OWSBackupImportItem new];
+        item.recordName = recordName;
+        item.encryptionKey = encryptionKey;
+        item.relativeFilePath = relativeFilePath;
+        item.uncompressedDataLength = uncompressedDataLength;
+        [items addObject:item];
+    }
+    return items;
+}
+
+- (void)downloadFilesFromCloud:(NSMutableArray<OWSBackupImportItem *> *)items
```

Comment:
TIOLI: download**Items**FromCloud?

Reply:
Yeah, naming was hard because you have one "file/item/record" per attachment.  Worse:

* I chose to use "recordName" for CloudKit record names, to emphasize that we were using these as CloudKit identifiers.
* Items are implementation details of the import and export process ... and export has two different kinds of item.

So, sometimes we deliberately download records _without_ downloading the file (e.g. when checking whether or not a record already existing) and other times we download record specifically to download the file.  I used this name here to emphasize that we were trying to do the latter.